Mathlib/MeasureTheory/Integral/SobolevInequality.lean

```diff
@@ -0,0 +1,619 @@
+compactly-supported function `u` on `ℝⁿ`, for `n ≥ 2`.  (More literally we encode `ℝⁿ` as
+`ι → ℝ` where `n := #ι` is finite and at least 2.)  Then the Lebesgue integral of the pointwise
+expression `|u x| ^ (n / (n - 1))` is bounded above by the `n / (n - 1)`-th power of the Lebesgue
+integral of the Fréchet derivative of `u`.
+
+For a basis-free version, see `lintegral_pow_le_pow_lintegral_fderiv`. -/
+theorem lintegral_pow_le_pow_lintegral_fderiv_aux
+    {p : ℝ} (hp : Real.IsConjExponent #ι p)
+    {u : (ι → ℝ) → F} (hu : ContDiff ℝ 1 u)
+    (h2u : HasCompactSupport u) :
+    ∫⁻ x, (‖u x‖₊ : ℝ≥0∞) ^ p ≤ (∫⁻ x, ‖fderiv ℝ u x‖₊) ^ p := by
+  have : (1:ℝ) ≤ ↑#ι - 1 := by
+    have hι : (2:ℝ) ≤ #ι := by exact_mod_cast Fintype.one_lt_card
+    linarith
+  calc ∫⁻ x, (‖u x‖₊ : ℝ≥0∞) ^ p
+      = ∫⁻ x, ((‖u x‖₊ : ℝ≥0∞) ^ (1 / (#ι - 1 : ℝ))) ^ (#ι : ℝ) := by
+        -- a little algebraic manipulation of the exponent
+        congr! 2 with x
+        rw [← ENNReal.rpow_mul, hp.conj_eq]
+        field_simp
+    _ = ∫⁻ x, ∏ _i : ι, (‖u x‖₊ : ℝ≥0∞) ^ (1 / (#ι - 1 : ℝ)) := by
+        -- express the left-hand integrand as a product of identical factors
+        congr! 2 with x
+        simp_rw [prod_const, card_univ]
+        norm_cast
+    _ ≤ ∫⁻ x, ∏ i, (∫⁻ xᵢ, ‖fderiv ℝ u (update x i xᵢ)‖₊) ^ ((1 : ℝ) / (#ι - 1 : ℝ)) := ?_
+    _ ≤ (∫⁻ x, ‖fderiv ℝ u x‖₊) ^ p :=
+        -- apply the grid-lines lemma
+        lintegral_prod_lintegral_pow_le _ hp (by fun_prop)
+  gcongr with x i
+  calc (‖u x‖₊ : ℝ≥0∞)
+      = (‖∫ xᵢ in Iic (x i), deriv (u ∘ update x i) xᵢ‖₊ : ℝ≥0∞) := by
+        -- apply the half-infinite fundamental theorem of calculus
+        have h3u : ContDiff ℝ 1 (u ∘ update x i) := hu.comp (by convert contDiff_update 1 x i)
+        have h4u : HasCompactSupport (u ∘ update x i) :=
+          h2u.comp_closedEmbedding (closedEmbedding_update x i)
+        simp [HasCompactSupport.integral_Iic_deriv_eq h3u h4u (x i)]
+    _ ≤ ∫⁻ xᵢ in Iic (x i), ‖deriv (u ∘ update x i) xᵢ‖₊ :=
+        ennnorm_integral_le_lintegral_ennnorm _ -- apply the triangle inequality
+    _ ≤ ∫⁻ xᵢ, (‖fderiv ℝ u (update x i xᵢ)‖₊ : ℝ≥0∞) := ?_
+  gcongr with y; swap; exact Measure.restrict_le_self
+  -- bound the derivative which appears
+  calc ‖deriv (u ∘ update x i) y‖₊ = ‖fderiv ℝ u (update x i y) (deriv (update x i) y)‖₊ := by
+        rw [fderiv.comp_deriv _ (hu.differentiable le_rfl).differentiableAt
+          (hasDerivAt_update x i y).differentiableAt]
+    _ ≤ ‖fderiv ℝ u (update x i y)‖₊ * ‖deriv (update x i) y‖₊ :=
+        ContinuousLinearMap.le_opNNNorm ..
+    _ ≤ ‖fderiv ℝ u (update x i y)‖₊ := by simp [deriv_update, Pi.nnnorm_single]
+
+variable {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E] [MeasurableSpace E] [BorelSpace E]
+  [FiniteDimensional ℝ E] (μ : Measure E) [IsAddHaarMeasure μ]
+
+open FiniteDimensional
+
+set_option linter.unusedVariables false in
```

Comment:
Why?

Reply:
It flags the hypotheses `hu` and `h2u` in the conclusion. But disabling the linter is not nice, I enabled it again.